Feat/#843 specimen exception endpoint

ciaranschutte opened the pull request:
**Description of changes**


- mostly moving functions and types into common modules / middleware for use in both program and entity exceptions
- moves parseTSV and upload to middlewares
- adds api, service, and repo calls for entity exception endpoints
- cleans up typing for clarity and reusability

**Type of Change**

- [ ] Bug
- [ ] Refactor
- [ ] New Feature
- [ ] Release Candidate

**Checklist before requesting review:**

- [ ] Check branch (code change PRs go to `develop` not master)
- [ ] Check copyrights for new files
- [ ] Manual testing
- [ ] Regression tests completed and passing (double check number of tests).
- [ ] Spelling has been checked.
- [ ] Updated swagger docs accordingly (check it's still valid)
- [ ] Set `validationDependency` in meta tag for [Argo Dictionary](https://github.com/icgc-argo/argo-dictionary) fields used in code

ciaranschutte commented on src/exception/types.ts:
```diff
@@ -65,6 +87,8 @@ export const isProgramExceptionRecord = (input: any): input is ProgramExceptionR
   );
 };
 
+export const isEntityExceptionRecord = (input: any): input is any => true;
```
this is extra validation

ciaranschutte replied:
leaving as is to not break parseTsv func contract

joneubank replied:
Does this do anything at all? This looks like placeholder code for actual implementation, if so there should at least be some comments explaining its purpose.

This actually seems quite dangerous as it will take any variable and make it `any` type, removing all type checking.

ciaranschutte commented:
mostly moving functions and types into common modules / middleware for use in both program and entity exceptions

joneubank commented:
Can you add description to the PR with summary of the work performed?

joneubank commented on src/exception/repo/entity.ts:
```diff
@@ -0,0 +1,70 @@
+        update,
+        { upsert: true, new: true },
+      ).lean(true);
+      // L.info(`doc created ${doc}`);
```
Remove unused commented code.

joneubank commented on src/routes/exception.ts:
```diff
@@ -19,19 +19,29 @@
 router.post(
   '/',
-  upload.single('programExceptionFile'),
+  parseTSV(isProgramExceptionRecord),
```
Using parseTSV here as middleware instead of in the service I think causes us to lose our type safety... see other comments

joneubank reviewed (changes requested):
Comments are mostly about the refactoring to move the parseTSV into express handlers which causes use to lose type checking... can we look at simplifying this so we don't end up passing around thigns typed as `any`?

ciaranschutte commented:
@joneubank 
- moved parseTSV back into the api function calls, using individual type guards to narrow `records` to specific types. Couldn't get this working with a single function like `(typeguard) =>  (validate as typeguard) => ...`

- updated `parseTsv` to use the global err handler

- rewrote `isEntityExceptionRecord` type guard instead of placeholder function

joneubank commented on src/exception/exception-api.ts:
```diff
@@ -41,39 +37,37 @@ function getResStatus(result: exceptionService.Result): number {
+  return records;
+};
+
+const validateEntityExceptionRecords = (records: ReadonlyArray<TsvUtils.TsvRecordAsJsonObj>) => {
```
To help future devs understand what this is doing, let's explicitly declare the return type:

```suggestion
const validateEntityExceptionRecords = (records: ReadonlyArray<TsvUtils.TsvRecordAsJsonObj>): ReadonlyArray<EntityExceptionRecord> => {
```

joneubank replied:
Same for `validateProgramExceptionRecords` on line 40

joneubank reviewed (commented):
Great updates, types are happy again!

joneubank reviewed (approved):
Thanks for all the chagnes!
